content/blog/2023-11-28-context-switching-sprs-on-powerpc.md

```diff
@@ -0,0 +1,443 @@
+	    old_thread->spefscr != new_thread->spefscr)
+		mtspr(SPRN_SPEFSCR, new_thread->spefscr);
+#endif
+#ifdef CONFIG_PPC_BOOK3S_64
+	if (cpu_has_feature(CPU_FTR_DSCR)) {
+		u64 dscr = get_paca()->dscr_default;
+		if (new_thread->dscr_inherit)
+			dscr = new_thread->dscr;
+
+		if (old_thread->dscr != dscr)
+			mtspr(SPRN_DSCR, dscr);
+	}
+
+	if (cpu_has_feature(CPU_FTR_ARCH_207S)) {
+		if (old_thread->bescr != new_thread->bescr)
+			mtspr(SPRN_BESCR, new_thread->bescr);
+		if (old_thread->ebbhr != new_thread->ebbhr)
+			mtspr(SPRN_EBBHR, new_thread->ebbhr);
+		if (old_thread->ebbrr != new_thread->ebbrr)
+			mtspr(SPRN_EBBRR, new_thread->ebbrr);
+
+		if (old_thread->fscr != new_thread->fscr)
+			mtspr(SPRN_FSCR, new_thread->fscr);
+
+		if (old_thread->tar != new_thread->tar)
+			mtspr(SPRN_TAR, new_thread->tar);
+	}
+
+	if (cpu_has_feature(CPU_FTR_P9_TIDR) &&
+	    old_thread->tidr != new_thread->tidr)
+		mtspr(SPRN_TIDR, new_thread->tidr);
+
+	if (cpu_has_feature(CPU_FTR_DEXCR_NPHIE) &&
+	    old_thread->hashkeyr != new_thread->hashkeyr)
+		mtspr(SPRN_HASHKEYR, new_thread->hashkeyr);
+#endif
+
+}
+```
+
+The gist is we first perform a series of `mfspr` operations, serialising the SPR
+values of the currently running task into its associated `task_struct`. Then we
+do a series of `mtspr` operations to deserialize the desired values of the new
+task back into the CPU.
+
+This procedure has two interesting optimisations, as explained by
+[the commit](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=v6.7-rc3&id=152d523e6307)
+that introduces `save_sprs()` and `restore_sprs()`:
+
+> powerpc: Create context switch helpers save_sprs() and restore_sprs()
+>
+> Move all our context switch SPR save and restore code into two helpers. We do
+> a few optimisations:
+>
+> - Group all mfsprs and all mtsprs. In many cases an mtspr sets a scoreboarding
+> bit that an mfspr waits on, so the current practise of mfspr A; mtspr A; mfpsr
+> B; mtspr B is the worst scheduling we can do.
+>
+> - SPR writes are slow, so check that the value is changing before writing it.
+
+And that's basically it, as far as the implementation goes at least. When first
+investigating this one question that kept nagging me was: why do we read these
+values here, instead of tracking them as they are set? The answer is that we
+either don't know what they were set to, or they could change themselves
+according to some event. As they are not general purpose read/write registers,
+and instead reflect some kind of CPU state, some SPRs will automatically update
+in response to events or changes in CPU state. For example, condition flags on
+arithmetic operations are reflected in the Condition Register (CR). Other times
+they might be directly accessible to userspace, which can set them arbitrarily
+without kernel knowledge. Some of these SPRs are also 'uninteresting' to the
+kernel, so it does not preserve them with the rest of the GPRs and highly used SPRs.
```

Comment:
is the list documented in the ABI or somewhere else?

Reply:
No, it's generally just SPRs the kernel doesn't change/use for itself for whatever reason. It wouldn't really make sense for the ABI to dictate that. Like the CTR vs the DEXCR; the kernel (nearly?) always changes the CTR, so needs to save it when taking a syscall. It doesn't change the DEXCR though unless swapping to a different task (as the userspace half might be different), so can defer saving it to when we actually do a swap.

I can add this example to the post.

Reply:
VRSAVE looks like a better example. Unprivileged, but the kernel only touches it to keep it correct across context switches. The kernel doesn't really care about it otherwise.

Reply:
Replaced that section anyway, it wasn't very clear and missed some possible points I think